tools/ncf.py

```diff
@@ -216,15 +223,45 @@ def parse_technique_methods(technique_file):
             args = [arg['value'].replace('\\"', '"') for arg in attribute['rval']['arguments']]
           if attribute['rval']['type'] == 'string':
             method_name = attribute['rval']['value']
+        # Extract class context from 'ifvarclass'
         elif attribute['lval'] == 'ifvarclass':
+          # Simple string get its value
           if attribute['rval']['type'] == 'string':
-            promise_class_context = class_context_and(class_context, attribute['rval']['value'])
+            ifvarclass_context = attribute['rval']['value']
+          # We have a function call here, and need to treat concat case
           if attribute['rval']['type'] == 'functionCall':
             ifvarclass_function = attribute['rval']['name']
-            ifvarclass_args = [arg['value'].replace('\\"', '"') for arg in attribute['rval']['arguments']]
-            promise_class_context = class_context_and(class_context, ifvarclass_function + '(' + ','.join(ifvarclass_args) + ')')
+            # Function is concat! We use that to handle variable in classes:
+            # variables in classes are expanded at runtime, making invalid character in classes
+            # We have to canonify variables only, and not the whole if var class
+            # as it would replace all the 'invalid' character from the class ( and '.' , not '!', ...)
+            # so a class like:
+            # "Monday.${bundle2.var}.debian.${bundle.var}.linux"
+            # will be written
+            # "concat('Monday.',canonify(${bundle2.var}),'.debian.',canonify(${bundle.var}),'.linux")
+            # But the class we really want to extract is:
+            # "Monday.${bundle2.var}.debian.${bundle.var}.linux"
+            if ifvarclass_function == 'concat':
+              ifvarclass_args = []
+              for arg in attribute['rval']['arguments']:
+                # simple string get only the value
+                if arg['type'] == 'string':
+                  ifvarclass_args.append(arg['value'])
+                # This a canonify call, extract only the value of the canonify
+                elif arg['type'] == 'functionCall' and arg['name'] == 'canonify':
+                  ifvarclass_args.append(arg['arguments'][0]['value'])
+                # Extract the function call correctly
+                else:
+                    function_call = parse_function_call_class_context(arg)
+                    ifvarclass_args.append(function_call)
+              ifvarclass_context = ''.join(ifvarclass_args)
+            # Another function call, extract it directly
+            else:
+              ifvarclass_context = parse_function_call_class_context(attribute['rval'])
             # FIXME: This is not really valid!
 
+      if ifvarclass_context:
```

Comment:
Please use "is not None" except if you want to catch the case where the variable is empty too.